src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -0,0 +1,356 @@
+      if (projectId && this.operationStartTime) {
+        const totalTime = endTime.getTime() - this.operationStartTime.getTime();
+        {
+          const operationAttributes = {
+            finalOperationStatus: info.finalOperationStatus,
+            streamingOperation: info.streamingOperation,
+            projectId,
+            instanceId: this.tabularApiSurface.instance.id,
+            table: this.tabularApiSurface.id,
+            cluster: this.cluster,
+            zone: this.zone,
+            appProfileId: this.tabularApiSurface.bigtable.appProfileId,
+            methodName: this.methodName,
+            clientName: `nodejs-bigtable/${version}`,
+          };
+          const metrics = {
+            operationLatency: totalTime,
+            retryCount: this.attemptCount - 1,
+            firstResponseLatency: this.firstResponseLatency ?? undefined,
+          };
+          this.metricsHandlers.forEach(metricsHandler => {
+            if (metricsHandler.onOperationComplete) {
+              metricsHandler.onOperationComplete(metrics, operationAttributes);
+            }
+          });
+        }
+      }
+    } else {
+      console.warn('Invalid state transition attempted');
+    }
+  }
+
+  /**
+   * Called when metadata is received. Extracts server timing information if available.
+   * @param {string} projectId The id of the project.
+   * @param {object} metadata The received metadata.
+   */
+  onMetadataReceived(
+    projectId: string,
+    metadata: {
+      internalRepr: Map<string, Buffer>;
+      options: {};
+    }
+  ) {
+    const mappedEntries = new Map(
+      Array.from(metadata.internalRepr.entries(), ([key, value]) => [
+        key,
+        value.toString(),
+      ])
+    );
+    const durationValues = mappedEntries.get('server-timing')?.split('dur=');
+    if (durationValues && durationValues[1]) {
+      if (!this.serverTimeRead) {
+        this.serverTimeRead = true;
+        const serverTime = parseInt(durationValues[1]);
+        if (projectId) {
+          this.serverTime = serverTime;
+        }
+      }
+    }
+  }
+
+  /**
+   * Called when status information is received. Extracts zone and cluster information.
+   * @param {object} status The received status information.
+   */
+  onStatusReceived(status: {
+    metadata: {internalRepr: Map<string, Buffer>; options: {}};
+  }) {
+    const mappedEntries = new Map(
+      Array.from(status.metadata.internalRepr.entries(), ([key, value]) => [
+        key,
+        value.toString(),
+      ])
+    );
+    const instanceInformation = mappedEntries
+      .get('x-goog-ext-425905942-bin')
+      ?.replace(new RegExp('\\n', 'g'), '')
+      .split(' \r'); // The data returned actually has a space after the zone.
+    if (instanceInformation && instanceInformation[0]) {
+      this.zone = instanceInformation[0];
+    }
+    if (instanceInformation && instanceInformation[1]) {
+      this.cluster = instanceInformation[1];
+    }
```

Comment:
I'm not super confident in this parsing logic. Are you sure these structure assumptions are safe? What if the cluster is present, but not zone? It seems like [the proto](https://github.com/googleapis/python-bigtable/blob/89b8da8a445aeb08854d9fa77cbc0e4fc042c87f/google/cloud/bigtable_v2/types/response_params.py#L31C7-L31C21) marks them both as optional

Reply:
The only way to be super confident in this parsing logic is to know what the server data looks like when there is a missing zone or missing cluster and I'm not really sure if there is any way to simulate that. Do you know if there is?

Reply:
It's not safe to assume that the server responses you get in testing are the only ones that will be returned.

You'll have to look at the spec, but my understanding is that the server returns a binary-encoded version of the [ResponseParams proto](https://github.com/googleapis/python-bigtable/blob/89b8da8a445aeb08854d9fa77cbc0e4fc042c87f/google/cloud/bigtable_v2/types/response_params.py#L31C7-L31C21). In Python, [we use ResponseParams.FromString](https://github.com/googleapis/python-bigtable/blob/b1d08c9662c92bd97bcb3e9b2fd0282b7faf38b1/google/cloud/bigtable/data/_metrics/data_model.py#L224) to convert it back into the proto object, and then read from the proto fields

Reply:
The custom parsing logic is gone now and we use ResponseParams.